foxglove/cmd/export.go

```diff
@@ -187,6 +188,572 @@ func validOutputFormat(format string) bool {
+		// request just received. Specifically, the timestamp of the last
+		// message written to the output file. By starting the request with this
+		// time, we will end up with some messages duplicated between the end of
+		// the previous file and the start of the next one. The merging process
+		// at the end will be in charge of accounting for these duplicates.
+		newStart := time.Unix(int64(info.maxTime)/1e9, int64(info.maxTime)%1e9)
+
+		if request.Start != nil && newStart == *request.Start {
+			repeatRequestCount++
+		} else {
+			repeatRequestCount = 0
+		}
+
+		// It is possible that the previous request contained a set of messagges
+		// on the same timestamp, right at the end of the response. In this
+		// instance, the next start time is the same as the previous start time.
+		// We can end up in a loop repeatedly writing the same messages out to
+		// partial files, and rerequesting the same data. To detect this we bail
+		// if we have requested the same data more than twice.
+		if repeatRequestCount > 1 {
+			debugf("got two successive requests with the same start time. Assuming EOF.")
+			break
+		}
+
+		request.Start = &newStart
+		if request.End == nil {
+			end := time.Now()
+			request.End = &end
+		}
+	}
+
+	// Now we need to combine the messages from the tmpfiles, handling the
+	// overlaps between them.
+
+	// If we have just one file, execute a mv. This will be the typical case
+	// when there is no failure.
+	if len(tmpfiles) == 1 {
+		debugf("single tmpfile - executing a rename")
+		err := os.Rename(tmpfiles[0].name, outputfile)
+		if err != nil {
+			return fmt.Errorf("failed to rename tmpfile: %w", err)
+		}
+		return nil
+	}
+	// otherwise go through the files in order and write out the messages
+	debugf("multiple tmpfiles - combining")
+	output, err := os.Create(outputfile)
+	if err != nil {
+		return err
+	}
+	defer output.Close()
+
+	switch request.OutputFormat {
+	case "bag1":
+		return combineBagTmpFiles(output, tmpfiles)
+	case "mcap0":
+		return combineMCAPTmpFiles(output, tmpfiles)
+	default:
+		return fmt.Errorf("unsupported format for resilient download: %s", request.OutputFormat)
+	}
+}
+
+func combineBagTmpFiles(w io.Writer, tmpfiles []partialFile) error {
+	debugf("combining %d bag files", len(tmpfiles))
+	writer, err := roslib.NewBagWriter(w)
+	if err != nil {
+		return fmt.Errorf("failed to construct output writer: %w", err)
+	}
+	connectionsWritten := make(map[uint32]bool)
+	var connIDIncrement, maxObservedConn uint32
+	for i, tmpfile := range tmpfiles {
+		if tmpfile.info.messageCount == 0 {
+			continue
+		}
+		_, err := tmpfile.rs.Seek(0, 0)
+		if err != nil {
+			return err
+		}
+		lexer, err := roslib.NewBagLexer(tmpfile.rs)
+		if err != nil {
+			return fmt.Errorf("failed to construct lexer: %w", err)
+		}
+		var scanThrough uint64
+		if i == len(tmpfiles)-1 {
+			scanThrough = tmpfile.info.maxTime
+		} else {
+			scanThrough = tmpfile.info.maxTime - 1
```

Comment:
this is where i see the underflow potentially happening, if the tempfile contains messages with timestamp 0.